modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2Tree.java

```diff
@@ -597,53 +601,96 @@ public void copyMetaInfo(MetaPageInfo info) throws IgniteCheckedException {
+                    inlinedCmpRes = inlineIdx.compare(pageAddr, off + fieldOff, inlineSize() - fieldOff, v2, comp);
 
-                    if (c == CANT_BE_COMPARE)
+                    if (inlinedCmpRes == CANT_BE_COMPARE)
                         break;
 
+                    if (inlineIdx instanceof StringInlineIndexColumn)
+                        resTrusted = false;
+
                     lastIdxUsed++;
 
-                    if (c != 0)
-                        return fixSort(c, inlineCols[i].sortType);
+                    if (inlinedCmpRes != 0) {
+                        if (!resTrusted)
+                            break;
+
+                        return fixSort(inlinedCmpRes, inlineCols[i].sortType);
+                    }
 
                     fieldOff += inlineIdx.fullSize(pageAddr, off + fieldOff);
 
                     if (fieldOff > inlineSize())
                         break;
                 }
 
-                if (lastIdxUsed == cols.length)
+                if (lastIdxUsed == cols.length && resTrusted)
                     return mvccCompare((H2RowLinkIO)io, pageAddr, idx, row);
 
                 inlineSizeRecomendation(row);
 
                 SearchRow rowData = getRow(io, pageAddr, idx);
 
-                for (int i = lastIdxUsed, len = cols.length; i < len; i++) {
-                    IndexColumn col = cols[i];
-                    int idx0 = col.column.getColumnId();
+                int dataRowCmpRes = 0;
 
-                    Value v2 = row.getValue(idx0);
+                if (!resTrusted)
+                    lastIdxUsed = 0;
 
-                    if (v2 == null) {
-                        // Can't compare further.
-                        return mvccCompare((H2RowLinkIO)io, pageAddr, idx, row);
-                    }
+                try {
+                    for (int i = lastIdxUsed, len = cols.length; i < len; i++) {
+                        IndexColumn col = cols[i];
+                        int idx0 = col.column.getColumnId();
+
+                        Value v2 = row.getValue(idx0);
+
+                        if (v2 == null) {
+                            // Can't compare further.
+                            return mvccCompare((H2RowLinkIO)io, pageAddr, idx, row);
+                        }
 
-                    Value v1 = rowData.getValue(idx0);
+                        Value v1 = rowData.getValue(idx0);
 
-                    int c = compareValues(v1, v2);
+                        dataRowCmpRes = compareValues(v1, v2);
 
-                    if (c != 0)
-                        return fixSort(c, col.sortType);
+                        if (dataRowCmpRes != 0)
+                            return fixSort(dataRowCmpRes, col.sortType);
+                    }
+                }
+                finally {
+                    if (!resTrusted && Integer.signum(inlinedCmpRes) != Integer.signum(dataRowCmpRes)) {
```

Comment:
inlinedCmpRes could have incorrect sign here. Suggestion above should fix the issue.

Reply:
it's OK cause `dataRowCmpRes` doesn't respect the sorting order either